### PR TITLE
Simplify core DPOR logic

### PR DESCRIPTION
**Pull request by gc00**
* Remove PROGRAM_TYPE, MC_TARGET, MC_SCHEDULER, etc.
  This is the first commit of several.  The goal is to simplify the core DPOR logic.
* Simplify core DPOR logic using mc_explore_branch()
  This defines mc_explore_branch() to replace both mc_initial_trace() and the inline code after `while (nextBranchPoint->hasValue())`
* Minor renaming and removing unneeded functions

### CODE REVIEW

File: src/mcmini_private.cpp [hunk 1]
```diff
@@ -200,59 +212,49 @@ mc_prepare_to_model_check_new_program()
...
+void
+mc_explore_branch(MCOptional<int> *nextBranchPoint)
 {
-  MC_PROGRAM_TYPE program = mc_fork_new_trace_at_main();
-  if (MC_IS_TARGET_PROGRAM(program)) return MC_TARGET_PROGRAM;
+  tid_t backtrackThread;
+
+  if (nextBranchPoint == nullptr /* initial branch */) {
+    mc_fork_new_trace();
+    backtrackThread = TID_MAIN_THREAD;
```
Comment (maxwellpirtle) [inline src/mcmini_private.cpp:223]: `MCOptional<>` attempt to act as `std::optional<T>`, the latter only having been introduced in C++17. Better is to pass the optional itself (instead of a pointer to the optional); check the `hasValue()` method; and then react appropriately.

File: src/mcmini_private.cpp [hunk 2]
```diff
@@ -200,59 +212,49 @@ mc_prepare_to_model_check_new_program()
...
-  // TODO: This idiom is fairly common... This simply allows
-  // the forked process to exit the constructor all of the stack
-  // frames find themselves in. Perhaps we could jump to the
-  // appropriate point with a getcontext()/setcontext() that will
-  // bring the trace process to the correct process as an alternative?
-  // It hurts readability to have the forked traces needing to escape
-  // in this way.
-  MC_PROGRAM_TYPE program = mc_run_initial_trace();
-  if (MC_IS_TARGET_PROGRAM(program)) return MC_TARGET_PROGRAM;
+  mc_explore_branch(nullptr); // nullptr means this is the initial branch
```
Comment (maxwellpirtle) [inline src/mcmini_private.cpp:250]: Here you would then pass `MCOptional<int>::nil()` instead of `nullptr`

File: src/mcmini_private.cpp [hunk 3]
```diff
@@ -200,59 +212,49 @@ mc_prepare_to_model_check_new_program()
...
   mc_exit_with_trace_if_necessary(traceId);
-  program = mc_rerun_current_trace_as_needed();
+  mc_run_next_trace_for_debugger();
+
   traceId++;
-  if (MC_IS_TARGET_PROGRAM(program)) return MC_TARGET_PROGRAM;
-  return MC_SCHEDULER;
+  if (nextBranchPoint != nullptr) { // if this is not the initial branch
+    *nextBranchPoint = programState->getDeepestDPORBranchPoint();
+  }
```
Comment (maxwellpirtle) [inline src/mcmini_private.cpp:242]: I'd move this `if` statement after what's now line 256 to give us something like:

```cpp
 while (nextBranchPoint.hasValue()) {
     MCOptional<int> *nextBranchPointPtr = &nextBranchPoint;
     mc_explore_branch(nextBranchPointPtr);
     *nextBranchPoint = programState->getDeepestDPORBranchPoint();
}
```

which will turn into (pending the other changes I suggested)
```cpp
 while (nextBranchPoint.hasValue()) {
     mc_explore_branch(nextBranchPoint.unwrapped());
     nextBranchPoint = programState->getDeepestDPORBranchPoint();
}
```
Moving the branch point over through modification of the pointer requires the reader to pick up on the fact that modification of the pointer of `nextBranchPointPtr` occurs inside this function. It is just a bit too subtle that's all: I  thought there was an issue on first glance at the changes made around the `while` loop

**Review by maxwellpirtle — CHANGES_REQUESTED** [review]
I need to take a closer look at some of the other changes more carefully before approving. In any case, there are some minor points to address in the interim

File: src/mcmini_private.cpp (hunk 1 above)
Comment (gc00) [reply src/mcmini_private.cpp:223]: I would prefer that the code be easily accessible by general programmers, rather than someone who knows the latest features of C++.  Both `MCOption<>~ and `std::optional<T>` will insert extra assembly code, and will cause GDB to refuse to show the value (because it's hidden in a new class).  Instead of asking people using GDB to write `nextBranchPoint ->MCOptionalXXXint.value`, it's a lot easier in GDB to just type `nextBranchPoint `.

Basically, I'd like to keep our C++ code close the actual pseudo-code, and not add extra layers that will be understood only by C++ mavens.  C++ is maybe the fourth most popular language after C, Java and Python.  And C++ is now facing competition from things like Go and Rust.  But if we stay close enough to C, then anybody will be able to read the code.

File: src/mcmini_private.cpp (hunk 2 above)
Comment (gc00) [reply src/mcmini_private.cpp:250]: And here, also, I would like to keep the code simple and easy to read.  Your change would make the code slightly slower.  And everyone will understand the current code, while the suggested code will appeal primarily to C++-enthusiasts.  I very much doubt that @lukagolf and @aayushi363 are in the habit of using these types of idioms.

Is there a reason to make this change, other than encouraging the reader of the code to go back and learn more about C++-specific idioms?

File: src/mcmini_private.cpp (hunk 3 above)
Comment (gc00) [reply src/mcmini_private.cpp:242]: As before, I think `nextBranchPointPtr != nullvalue` is clearer to a general reader of the code than `nextBranchPoint.hasValue()`.  Note also that GDB will not allow you to look inside `nextBranchPoint.hasValue()` to see what it means.  Let's keep the code easy for a general programmer, as opposed to a C++ enthusiast.

**Comment by gc00** [comment]
@maxwellpirtle ,
    To summarize the issues above, it all comes from the fact that nextBranchPoint is actually an 'int'.  But the current code disguises the fact that it is an 'int'.  An even better version of this code would be to replace:
```
246       mc_explore_branch(nullptr); // nullptr means this is the initial branch
247   
248       MCOptional<int> nextBranchPoint = programState->getDeepestDPORBranchPoint();
249       while (nextBranchPoint.hasValue()) {
250         MCOptional<int> *nextBranchPointPtr = &nextBranchPoint;
251         mc_explore_branch(nextBranchPointPtr);
252         nextBranchPoint = programState->getDeepestDPORBranchPoint();
253       }  
```
with:
```
#define FIRST_BRANCH (-1)
249       while (nextBranchPoint != FIRST_BRANCH) {
251         mc_explore_branch(nextBranchPoint);
252         nextBranchPoint = programState->getDeepestDPORBranchPoint();
253       } 
```
  (where `MCState::getDeepestDPORBranchPoint()`` is modified to return FIRST_BRANCH instead of nullptr,
    
This has the further benefit of no longer needing the obscure line, below, to convert from 'int' to 'int':
```
                                      const int bp = nextBranchPoint.unwrapped();  
```
which has its own 8-line definition of `MCState.h:unwrapped()` as an inline function in MCState.h.

So, in the end, this allows us to reduce the current `mc_do_model_checking()` from about 30 lines of code down to about 4 lines of code, as I ilustrated above.  I argue that the 4 lines of code is both shorter than the 30 lines of code, and also __easier__ to understand than the original 30 lines of code.  Once this current PR is accepted, I can then reduce it again -- this time down to the 4 lines of code illustrated above, and separately removing unnecessary inline methods, such as `unwrapped()` in MCState.h.

File: src/mcmini_private.cpp (hunk 1 above)
Comment (maxwellpirtle) [reply src/mcmini_private.cpp:223]: The point here is that you should either use `MCOptional<>` or `int*`, but don't combine the two. The concept of an optional value should be familiar enough to most students to be accessible in my opinion (see my larger comment below)

**Comment by maxwellpirtle** [comment]
I agree that debugging is easier in C than with C++ and GDB gets confused more easily with the latter than with the former

Ultimately, if we're trying to move `mcmini` more towards `C` than `C++`, bigger structural changes would be needed since a lot of its current structure is based around object-oriented C++ concepts rather than the functional approach taken by `C`. We originally had `mcmini` entirely in C and moved to C++ since we were looking both for a) implementations of standard data structures and b) the ability to make mcmini extensible. Such extensibility is certainly possible with `C`, but since my experience is primarily with object-oriented paradigms I was better able to express such extensibility with `C++`. Of course, this may come back to bite us as we continue to add more low-level features, especially if we want to allow for plugins into McMini since `C++` name mangling is rather prohibitive when it comes to finding the right symbols. Likewise, the inheritance scheme in `C++` makes it a bit tricky with dynamic libraries and the `LD_PRELOAD` business we're playing currently.

File: src/mcmini_private.cpp (hunk 2 above)
Comment (maxwellpirtle) [reply src/mcmini_private.cpp:250]: As I mentioned above, the concept of an optional value should be familiar. All we're trying to do is add that into C++11 since `std::optional<T>` doesn't exist until later. In terms of speed, sure it could be slower sure. Again, if we're looking to move into a `C`-like codebase, we'd need to make bigger changes ultimately

**Comment by gc00** [comment]
@maxwellpirtle , My argument is not with object-oriented programming.  The changes proposed here still use object-oriented programming.  I just want it to be easy to read for a non-expert in C++.  Let's talk 1-on-1.

**Comment by maxwellpirtle** [comment]
@gc00 just [use either `nullptr` or `MCOptional<int>` but not both](https://github.com/mcminickpt/mcmini/pull/83#discussion_r1207834651) and I'll approve this PR

**Comment by gc00** [comment]
@maxwellpirtle ,
    Got it.  Thanks!   I'll use nullptr only.  (I still had `MCOptional<int>` there because I wanted to make several small, incremental changes only.  But ultimately, I was to stop using `MCOptional<int>` here.)
    I'm away for a long weekend, and I'll make this change afterward.

**Comment by gc00** [comment]
@maxwellpirtle , I've updated this PR to make the code still simpler, based on using the constant FIRST_BRANCH, as described earlier in the conversation.  I hope we have finally converged.  :-)

**Review by maxwellpirtle — APPROVED** [review]
This looks good to me. The only thing we should probably look into in the future is the possibility that `getcontext()/setcontext()` doesn't destroy C++ objects on the stack.

**Comment by gc00** [comment]
@maxwellpirtle , Excellent!  With this in place, it will make things simpler -- both for deep debugging, and for me to get back and finish debugging a bit of the GDB interface that needs to be fixed.

And next, we should get back to PR #85.  But that is a matter of renaming that, in hindsight, reflect better the variables and function.  But that is not as urgent as this refactoring of the core logic.
